### PR TITLE
feat(wall-x): add language runtime and text supervision

### PR DESCRIPTION
**Pull request by pkooij**
## Summary

Add WALL-X language generation and recipe-supervised training on top of the minimal shared contract in #4183.

After the restack, this PR contains only WALL-X integration: 1 commit, 6 files, +914 / -266. The generic processor-rebuild fix now lives in #4183, and no rollout or shared runtime code is duplicated here.

## Runtime

- Put RenderMessagesStep first in the normal WALL-X input pipeline.
- Preserve WALL-X native Qwen chat formatting, camera/image placement, multimodal processing, and tokenization in its serializable processor.
- Implement direct supports_text_generation and generate_text behavior for VQA and next-subtask requests.
- Feed a generated subtask back through the normal task field for action inference.
- Emit fast-action generation prompt token IDs from the processor, so policy methods consume model-ready inputs without retokenizing caller text.

## Training

- Store the built-in WALL-OSS wording as a policy-local recipe with optional YAML override.
- Apply sparse next-token cross-entropy only to selected assistant targets, including the closing chat token.
- Support action-only, text-only, and joint text-plus-flow rows.
- Flatten say tool calls to say-tag supervision and exclude prompts, images, state, action-query tokens, and padding from text loss.
- Keep flow and text loss weights configurable.

## Compatibility

There is intentionally no legacy processor reconciliation shim. Language-enabled checkpoints must save or recreate the current WALL-X processor containing RenderMessagesStep. The resolved recipe is stored inline in config for portable reloads.

Updated checkpoint: https://huggingface.co/pepijn223/wallx-super-chatton-subtask/commit/1178dbdbe16f98ed6e6fb65131855051ef3e7cb5

## Validation

- WALL-X plus shared contract tests: 20 passed, 2 skipped
- Pre-commit hooks pass on every touched file

### CODE REVIEW

File: docs/source/walloss.mdx [hunk 1]
```diff
@@ -59,6 +98,33 @@ lerobot-train \
...
+
+WALL-X includes a built-in recipe for language training. Set `recipe_path`
+to override it when training from LeRobot's optional language columns:
+
+```bash
+lerobot-train \
+  --dataset.repo_id=your_org/your_language_annotated_dataset \
+  --policy.type=wall_x \
+  --policy.pretrained_name_or_path=x-square-robot/wall-oss-flow \
+  --policy.recipe_path=subtask_joint.yaml \
```
Comment (imstevenpmwork) [inline docs/source/walloss.mdx]: interesting that we need this but eo1 doesn't

File: docs/source/walloss.mdx [hunk 2]
```diff
@@ -59,6 +98,33 @@ lerobot-train \
...
+  --policy.device=cuda \
+  --batch_size=1 \
+  --steps=20000
+```
+
+Targeted assistant turns receive next-token cross-entropy, while `low_level`
+recipe streams receive continuous flow-action supervision. A row may receive
+either objective or both. User prompts, images, proprioception, action-query
+tokens, and padding are excluded from text loss. With `policy.recipe=null`,
+WALL-OSS keeps its original action-only training path.
```
Comment (imstevenpmwork) [inline docs/source/walloss.mdx]: Apparently without it, we wouldn't train on language

File: src/lerobot/policies/wall_x/configuration_wall_x.py [hunk 3]
```diff
@@ -66,6 +91,18 @@ class WallXConfig(PreTrainedConfig):
     # otherwise falls back to the native per-chunk SDPA implementation.
     vision_attn_implementation: str = "auto"
 
+    # Optional explicit external language-recipe override.
+    recipe_path: str | None = None
+    # WALL-X's language contract: defaults to the WALL-OSS trained subtask wording;
+    # a fine-tune with `recipe_path` replaces it, and the checkpoint then prompts
+    # itself with the recipe it was trained on.
+    recipe: dict | None = field(default_factory=_wall_x_default_recipe)
```
Comment (imstevenpmwork) [inline src/lerobot/policies/wall_x/configuration_wall_x.py:99]: seems is only outdate docs then

File: src/lerobot/policies/wall_x/configuration_wall_x.py [hunk 4]
```diff
@@ -80,6 +117,11 @@ class WallXConfig(PreTrainedConfig):
     def __post_init__(self):
         super().__post_init__()
 
+        if self.recipe_path is not None:
+            from lerobot.datasets.recipe import resolve_recipe_override
+
+            self.recipe = asdict(resolve_recipe_override(self.recipe, self.recipe_path))
```
Comment (imstevenpmwork) [inline src/lerobot/policies/wall_x/configuration_wall_x.py:123]: we should do a follow-up pr to move language & recipe related stuff from datasets to its own module

File: src/lerobot/policies/wall_x/modeling_wall_x.py [hunk 5]
```diff
@@ -2056,16 +1932,34 @@ def forward(self, batch: dict[str, Tensor]) -> tuple[Tensor, dict]:
         Returns:
             tuple: (loss, loss_dict)
         """
-        batch = self.preprocess_inputs(batch, compute_position_ids=True)
+        recipe_supervision = "messages_rendered" in batch
```
Comment (imstevenpmwork) [inline src/lerobot/policies/wall_x/modeling_wall_x.py]: use constant

File: src/lerobot/policies/wall_x/processor_wall_x.py [hunk 6]
```diff
@@ -112,3 +144,290 @@ def transform_features(
...
+        processor, action_tokenizer = self._get_processors()
+        from .constant import GENERATE_SUBTASK_RATIO, MODEL_TYPE, PRIORITY_ORDER
+        from .modeling_wall_x import _prepare_wall_x_image_inputs
+        from .utils import (
+            get_wallx_normal_text,
+            img_key_mapping,
+            preprocesser_call,
+            process_grounding_points,
+            replace_action_token,
+        )
```
Comment (imstevenpmwork) [inline src/lerobot/policies/wall_x/processor_wall_x.py]: move imports to top

File: src/lerobot/policies/wall_x/processor_wall_x.py [hunk 7]
```diff
@@ -112,3 +144,290 @@ def transform_features(
         self, features: dict[PipelineFeatureType, dict[str, PolicyFeature]]
     ) -> dict[PipelineFeatureType, dict[str, PolicyFeature]]:
         return features
+
+
+@dataclass
+@ProcessorStepRegistry.register(name="wall_x_tokenizer")
+class WallXTokenizerStep(ProcessorStep):
```
Comment (imstevenpmwork) [inline src/lerobot/policies/wall_x/processor_wall_x.py]: it almost feels like this can be spliced in two to use the specialized processors

File: src/lerobot/policies/wall_x/utils.py [hunk 8]
```diff
@@ -42,6 +42,9 @@
     CAMERA_NAME_MAPPING,
 )
 
+TEXT_TARGET_START = "<|wall_text_target_start|>"
+TEXT_TARGET_END = "<|wall_text_target_end|>"
+
 
 @dataclass
 class X2RDataProcessingConfig:
```
Comment (imstevenpmwork) [inline src/lerobot/policies/wall_x/utils.py]: wonder why this doesn't live in the config file instead

File: src/lerobot/policies/wall_x/utils.py [hunk 9]
```diff
@@ -279,6 +299,35 @@ def preprocesser_call(
     return BatchFeature(data={**text_inputs, **image_inputs, **videos_inputs})
 
 
+def _extract_text_target_spans(text: str) -> tuple[str, list[tuple[int, int]]]:
+    clean_parts: list[str] = []
+    spans: list[tuple[int, int]] = []
+    cursor = 0
+    clean_length = 0
+    while True:
```
Comment (imstevenpmwork) [inline src/lerobot/policies/wall_x/utils.py]: smelly

File: docs/source/walloss.mdx (hunk 1 above)
Comment (pkooij) [reply docs/source/walloss.mdx]: It's not needed but can be overwritten. I will remove the other yaml recipes in a later PR so that only place to define recipes will be the config itself

File: src/lerobot/policies/wall_x/configuration_wall_x.py (hunk 4 above)
Comment (pkooij) [reply src/lerobot/policies/wall_x/configuration_wall_x.py:123]: yes agree

File: src/lerobot/policies/wall_x/processor_wall_x.py (hunk 7 above)
Comment (pkooij) [reply src/lerobot/policies/wall_x/processor_wall_x.py]: I split them in prompt rendering and tokenization here https://github.com/huggingface/lerobot/pull/4234/changes/ecb9dde59a1addadd51f8b2c200a5099307555a2#diff-75ab10be799b77e3c1b9cd30404666b3e869a9ee6779b723382be1f7fdf287f8

File: src/lerobot/policies/wall_x/processor_wall_x.py (hunk 6 above)
Comment (pkooij) [reply src/lerobot/policies/wall_x/processor_wall_x.py]: done in https://github.com/huggingface/lerobot/pull/4234/changes/ecb9dde59a1addadd51f8b2c200a5099307555a2#diff-75ab10be799b77e3c1b9cd30404666b3e869a9ee6779b723382be1f7fdf287f8

File: src/lerobot/policies/wall_x/modeling_wall_x.py (hunk 5 above)
Comment (pkooij) [reply src/lerobot/policies/wall_x/modeling_wall_x.py]: done in https://github.com/huggingface/lerobot/pull/4234/changes/ecb9dde59a1addadd51f8b2c200a5099307555a2#diff-75ab10be799b77e3c1b9cd30404666b3e869a9ee6779b723382be1f7fdf287f8

File: src/lerobot/policies/wall_x/configuration_wall_x.py (hunk 3 above)
Comment (pkooij) [reply src/lerobot/policies/wall_x/configuration_wall_x.py:99]: updates the doc in https://github.com/huggingface/lerobot/pull/4234/changes/ecb9dde59a1addadd51f8b2c200a5099307555a2#diff-75ab10be799b77e3c1b9cd30404666b3e869a9ee6779b723382be1f7fdf287f8

File: docs/source/walloss.mdx (hunk 2 above)
Comment (pkooij) [reply docs/source/walloss.mdx]: No its outdated, updated doc in https://github.com/huggingface/lerobot/pull/4234/changes/ecb9dde59a1addadd51f8b2c200a5099307555a2#diff-75ab10be799b77e3c1b9cd30404666b3e869a9ee6779b723382be1f7fdf287f8

File: src/lerobot/policies/wall_x/utils.py (hunk 9 above)
Comment (pkooij) [reply src/lerobot/policies/wall_x/utils.py]: removed and modified it in https://github.com/huggingface/lerobot/pull/4234/changes/ecb9dde59a1addadd51f8b2c200a5099307555a2

File: src/lerobot/policies/wall_x/utils.py (hunk 8 above)
Comment (pkooij) [reply src/lerobot/policies/wall_x/utils.py]: found out that class is dead code actually, removed it in https://github.com/huggingface/lerobot/pull/4234/changes/060fb46c2d942fccb2b027cc1736efee3cbae2e6

**Comment by imstevenpmwork** [comment]
@claude Review this PR. Also, is there any dead code introduced ?

**Review by imstevenpmwork — DISMISSED** [review]
LGTM, thanks. I kicked Claude for curiosity

**Comment by pkooij** [comment]
@claude are your previous comments addressed?
